python_streams/streams.py

```diff
@@ -73,6 +170,17 @@ def append(self, item: T) -> 'Stream[T]':
     def extend(self, item: T) -> 'Stream[T]':
         return self.chain(Stream((item,)))
 
-    @lru_cache(1)
+    class AlreadyConsumed(Exception):
+        pass
+
     def to_list(self) -> List[T]:
-        return list(self.items)
+        if self.with_cache:
```

Comment:
Can we make self.items be an `@property` that returns either the list if with_cache=True and cache is not None and otherwise returns the generator (which is currently self.items, we can rename it to self._generator)?

Reply:
The nice thing about this is that if we do that then every method in this class automatically uses the cache if it's available instead of throwing an exception, without changing any code.